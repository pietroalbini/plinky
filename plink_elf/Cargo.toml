[package]
name = "plink_elf"
version = "0.1.0"
edition = "2021"

[dependencies]
plink_macros = { path = "../plink_macros" }

[dev-dependencies]
anyhow = "1.0.71"
insta = "1.30.0"
tempfile = "3.6.0"
