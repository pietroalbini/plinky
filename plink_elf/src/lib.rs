pub mod errors;
mod reader;
mod types;
mod utils;
mod writer;

pub use self::types::*;
