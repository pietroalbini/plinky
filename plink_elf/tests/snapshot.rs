//! Check whether ELF files are parsed correctly.

use anyhow::Error;
use plink_elf::Object;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use std::process::Command;
use tempfile::NamedTempFile;

macro_rules! test {
    ($name:ident, $source:expr $(, $variant:ident)*) => {
        mod $name {
            use super::*;

            $(
                #[test]
                fn $variant() -> Result<(), Error> {
                    implement_test($source, stringify!($variant))
                }
            )*
        }
    };
}

test!(
    hello_asm,
    "tests/snapshot/hello_asm.S",
    x86,
    x86_64,
    x86_linked,
    x86_64_linked
);
test!(hello_c, "tests/snapshot/hello_c.c", x86, x86_64);

#[track_caller]
fn implement_test(source: &str, variant: &str) -> Result<(), Error> {
    let (variant, link) = match variant {
        "x86" => (Variant::X86, Link::No),
        "x86_64" => (Variant::X86_64, Link::No),
        "x86_linked" => (Variant::X86, Link::Yes),
        "x86_64_linked" => (Variant::X86_64, Link::Yes),
        other => panic!("unsupported variant: {other}"),
    };

    let mut object_file = match source.rsplit_once('.').map(|(_name, ext)| ext) {
        Some("S") => compile_asm(source, variant)?,
        Some("c") => compile_c(source, variant)?,
        Some(other) => panic!("unsupported extension: {other}"),
        None => panic!("missing extension for {source}"),
    };
    if let Link::Yes = link {
        object_file = link_single_object(object_file.path(), variant)?;
    }
    let mut file = BufReader::new(File::open(object_file.path())?);

    let parsed = Object::load(&mut file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_omit_expression(true);
    settings.set_snapshot_path("snapshot");
    settings.set_prepend_module_to_snapshot(false);
    settings.set_snapshot_suffix(match (variant, link) {
        (Variant::X86, Link::No) => "32bit",
        (Variant::X86_64, Link::No) => "64bit",
        (Variant::X86, Link::Yes) => "32bit-linked",
        (Variant::X86_64, Link::Yes) => "64bit-linked",
    });
    let _guard = settings.bind_to_scope();

    let name = source
        .rsplit_once('/')
        .map(|(_dir, name)| name)
        .unwrap_or(source);
    let name = name
        .rsplit_once('.')
        .map(|(name, _ext)| name)
        .unwrap_or(name);

    insta::assert_snapshot!(name, format!("{parsed:#x?}"));
    Ok(())
}

fn compile_asm(source: &str, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("nasm")
        .arg("-o")
        .arg(dest.path())
        .arg("-f")
        .arg(match variant {
            Variant::X86 => "elf32",
            Variant::X86_64 => "elf64",
        })
        .arg(source)
        .status()?;
    anyhow::ensure!(status.success(), "failed to compile {}", source);
    Ok(dest)
}

fn compile_c(source: &str, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("gcc")
        .arg("-c")
        .arg("-o")
        .arg(dest.path())
        .arg(source)
        .arg(match variant {
            Variant::X86 => "-m32",
            Variant::X86_64 => "-m64",
        })
        .status()?;
    anyhow::ensure!(status.success(), "failed to compile {}", source);
    Ok(dest)
}

fn link_single_object(object: &Path, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("ld")
        .arg("-o")
        .arg(dest.path())
        .arg(object)
        .args(match variant {
            Variant::X86 => ["-m", "elf_i386"],
            Variant::X86_64 => ["-m", "elf_x86_64"],
        })
        .status()?;
    anyhow::ensure!(status.success(), "failed to link {object:?}");
    Ok(dest)
}

#[derive(Clone, Copy)]
enum Variant {
    X86,
    X86_64,
}

#[derive(Clone, Copy)]
enum Link {
    Yes,
    No,
}
