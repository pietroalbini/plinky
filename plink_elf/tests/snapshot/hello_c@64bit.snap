---
source: plink_elf/tests/snapshot.rs
---
ElfObject {
    env: ElfEnvironment {
        class: Elf64,
        endian: Little,
        abi: SystemV,
        machine: X86_64,
    },
    type_: Relocatable,
    entry: None,
    flags: 0x0,
    sections: {
        "#00 ": ElfSection {
            name: "",
            memory_address: 0x0,
            content: Null,
        },
        "#01 .text": ElfSection {
            name: ".text",
            memory_address: 0x0,
            content: Program(
                ElfProgramSection {
                    perms: ElfPermissions(RX),
                    raw: RawBytes {
                             ?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?a ?b ?c ?d ?e ?f   (len: 0x29)
                           +-------------------------------------------------+
                        0? | f3 0f 1e fa 55 48 89 e5 48 83 ec 10 89 7d fc 48 | ....UH..H....}.H
                        1? | 89 75 f0 48 8d 05 00 00 00 00 48 89 c7 e8 00 00 | .u.H......H.....
                        2? | 00 00 b8 00 00 00 00 c9 c3                      | .........
                    },
                },
            ),
        },
        "#02 .rela.text": ElfSection {
            name: ".rela.text",
            memory_address: 0x0,
            content: RelocationsTable(
                ElfRelocationsTable {
                    symbol_table: "#11 .symtab",
                    applies_to_section: "#01 .text",
                    relocations: [
                        ElfRelocation {
                            offset: 0x16,
                            symbol: "#3 <empty>",
                            relocation_type: X86_64_PC32,
                            addend: Some(
                                0xfffffffffffffffc,
                            ),
                        },
                        ElfRelocation {
                            offset: 0x1e,
                            symbol: "#5 puts",
                            relocation_type: X86_64_PLT32,
                            addend: Some(
                                0xfffffffffffffffc,
                            ),
                        },
                    ],
                },
            ),
        },
        "#03 .data": ElfSection {
            name: ".data",
            memory_address: 0x0,
            content: Program(
                ElfProgramSection {
                    perms: ElfPermissions(RW),
                    raw: RawBytes {},
                },
            ),
        },
        "#04 .bss": ElfSection {
            name: ".bss",
            memory_address: 0x0,
            content: Unknown(
                ElfUnknownSection {
                    id: 0x8,
                    raw: RawBytes {},
                },
            ),
        },
        "#05 .rodata": ElfSection {
            name: ".rodata",
            memory_address: 0x0,
            content: Program(
                ElfProgramSection {
                    perms: ElfPermissions(R),
                    raw: RawBytes {
                             ?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?a ?b   (len: 0xc)
                           +-------------------------------------+
                        0? | 48 65 6c 6c 6f 20 77 6f 72 6c 64 00 | Hello world.
                    },
                },
            ),
        },
        "#06 .comment": ElfSection {
            name: ".comment",
            memory_address: 0x0,
            content: Program(
                ElfProgramSection {
                    perms: ElfPermissions(),
                    raw: RawBytes {
                             ?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?a ?b ?c ?d ?e ?f   (len: 0x2c)
                           +-------------------------------------------------+
                        0? | 00 47 43 43 3a 20 28 55 62 75 6e 74 75 20 31 31 | .GCC: (Ubuntu 11
                        1? | 2e 34 2e 30 2d 31 75 62 75 6e 74 75 31 7e 32 32 | .4.0-1ubuntu1~22
                        2? | 2e 30 34 29 20 31 31 2e 34 2e 30 00             | .04) 11.4.0.
                    },
                },
            ),
        },
        "#07 .note.GNU-stack": ElfSection {
            name: ".note.GNU-stack",
            memory_address: 0x0,
            content: Program(
                ElfProgramSection {
                    perms: ElfPermissions(),
                    raw: RawBytes {},
                },
            ),
        },
        "#08 .note.gnu.property": ElfSection {
            name: ".note.gnu.property",
            memory_address: 0x0,
            content: Note(
                ElfNotesTable {
                    notes: [
                        ElfNote {
                            name: "GNU",
                            value: RawBytes {
                                     ?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?a ?b ?c ?d ?e ?f   (len: 0x10)
                                   +-------------------------------------------------+
                                0? | 02 00 00 c0 04 00 00 00 03 00 00 00 00 00 00 00 | ................
                            },
                            type_: 0x5,
                        },
                    ],
                },
            ),
        },
        "#09 .eh_frame": ElfSection {
            name: ".eh_frame",
            memory_address: 0x0,
            content: Program(
                ElfProgramSection {
                    perms: ElfPermissions(R),
                    raw: RawBytes {
                             ?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?a ?b ?c ?d ?e ?f   (len: 0x38)
                           +-------------------------------------------------+
                        0? | 14 00 00 00 00 00 00 00 01 7a 52 00 01 78 10 01 | .........zR..x..
                        1? | 1b 0c 07 08 90 01 00 00 1c 00 00 00 1c 00 00 00 | ................
                        2? | 00 00 00 00 29 00 00 00 00 45 0e 10 86 02 43 0d | ....)....E....C.
                        3? | 06 60 0c 07 08 00 00 00                         | .`......
                    },
                },
            ),
        },
        "#10 .rela.eh_frame": ElfSection {
            name: ".rela.eh_frame",
            memory_address: 0x0,
            content: RelocationsTable(
                ElfRelocationsTable {
                    symbol_table: "#11 .symtab",
                    applies_to_section: "#09 .eh_frame",
                    relocations: [
                        ElfRelocation {
                            offset: 0x20,
                            symbol: "#2 <empty>",
                            relocation_type: X86_64_PC32,
                            addend: Some(
                                0x0,
                            ),
                        },
                    ],
                },
            ),
        },
        "#11 .symtab": ElfSection {
            name: ".symtab",
            memory_address: 0x0,
            content: SymbolTable(
                ElfSymbolTable {
                    symbols: {
                        "#0 <empty>": ElfSymbol {
                            name: "",
                            binding: Local,
                            type_: NoType,
                            definition: Undefined,
                            value: 0x0,
                            size: 0x0,
                        },
                        "#1 hello_c.c": ElfSymbol {
                            name: "hello_c.c",
                            binding: Local,
                            type_: File,
                            definition: Absolute,
                            value: 0x0,
                            size: 0x0,
                        },
                        "#2 <empty>": ElfSymbol {
                            name: "",
                            binding: Local,
                            type_: Section,
                            definition: Section(
                                "#01 .text",
                            ),
                            value: 0x0,
                            size: 0x0,
                        },
                        "#3 <empty>": ElfSymbol {
                            name: "",
                            binding: Local,
                            type_: Section,
                            definition: Section(
                                "#05 .rodata",
                            ),
                            value: 0x0,
                            size: 0x0,
                        },
                        "#4 main": ElfSymbol {
                            name: "main",
                            binding: Global,
                            type_: Function,
                            definition: Section(
                                "#01 .text",
                            ),
                            value: 0x0,
                            size: 0x29,
                        },
                        "#5 puts": ElfSymbol {
                            name: "puts",
                            binding: Global,
                            type_: NoType,
                            definition: Undefined,
                            value: 0x0,
                            size: 0x0,
                        },
                    },
                },
            ),
        },
        "#12 .strtab": ElfSection {
            name: ".strtab",
            memory_address: 0x0,
            content: StringTable(
                ElfStringTable {
                    strings: {
                        0x0: "",
                        0x1: "hello_c.c",
                        0xb: "main",
                        0x10: "puts",
                    },
                },
            ),
        },
        "#13 .shstrtab": ElfSection {
            name: ".shstrtab",
            memory_address: 0x0,
            content: StringTable(
                ElfStringTable {
                    strings: {
                        0x0: "",
                        0x1: ".symtab",
                        0x9: ".strtab",
                        0x11: ".shstrtab",
                        0x1b: ".rela.text",
                        0x26: ".data",
                        0x2c: ".bss",
                        0x31: ".rodata",
                        0x39: ".comment",
                        0x42: ".note.GNU-stack",
                        0x52: ".note.gnu.property",
                        0x65: ".rela.eh_frame",
                    },
                },
            ),
        },
    },
    segments: [],
}
